Sources/EventHub/EventHub.swift

```diff
@@ -0,0 +1,213 @@
+        
+        DispatchQueue.main.asyncAfter(deadline: .now() + pingTimeout) { [weak self] in
+            self?.sendPing()
+        }
+    }
+    
+    private func getNextRequestId() -> UInt64 {
+        lastRequestIdLock.lock()
+        defer { lastRequestIdLock.unlock() }
+        lastRequestId += 1
+        return lastRequestId
+    }
+    
+    private func disconnect() {
+        status = .disconnected
+        ws.disconnect()
+        pendingPings = [:]
+    }
+    
+    private func reconnectAfter(delay: TimeInterval) {
+        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
+            self?.reconnect()
+        }
+    }
+    
+    private func removeSubscribtion(at id: UInt64) {
+        if let subscribtion = subscribtions[id] {
+            let requestId = getNextRequestId()
+            ws.send(data: encoder.encodeUnsubscribe(requestId: requestId, topic: subscribtion.topic))
+            
+            subscribtions.removeValue(forKey: id)
+            if subscribtions.count == 0 {
+                disconnect()
+            }
+        }
+    }
+
+    private func onReceive(data: Data) {
+        guard let id = decoder.decodeResponseId(data: data) else {
+            // something unexpected
+            return
+        }
+        
+        if let subscribtion = subscribtions[id] {
+            guard let message = decoder.decodeMessage(data: data) else {
+                // probably a subscribtion response
+                return
+            }
+            
+            subscribtion.subject.send(message)
+            return
+        }
+        
+        if pendingPings.keys.contains(id) {
+            pendingPings.removeValue(forKey: id)
+            return
+        }
+    }
+    
+    public func pause() {
+        disconnect()
+        status = .paused
+    }
+    
+    public func resume() {
+        if status == .paused {
+            status = .disconnected
+        }
+        
+        if subscribtions.count > 0 {
+            reconnect()
+        }
+    }
+    
+    public func subscribe(topic: String) -> AnyPublisher<MessageResponse, Never> {
+        reconnect()
+        
+        let subscribtion: Subscribtion
+        
+        if let existingSubscribtion = subscribtions.values.first(where: { $0.topic == topic }) {
+            subscribtion = Subscribtion(existingSubscribtion,
+                                        subscribersCount: existingSubscribtion.subscribersCount + 1)
+        } else {
+            let requestId = getNextRequestId()
+            subscribtion = Subscribtion(id: requestId, topic: topic)
+            if (status == .connected) {
+                ws.send(data: encoder.encodeSubscribe(requestId: requestId, topic: topic))
+            }
+        }
+        
+        subscribtions[subscribtion.id] = subscribtion
+        
+        return subscribtion.subject.handleEvents(receiveCancel: {
+            if let canceledSubscribtion = self.subscribtions[subscribtion.id] {
```

Comment:
[weak self] is probably not necessary here, but I'd still add it just in case. I think you'll have a strong reference to your instance of `EventHub` anyway and that would keep it alive.